### PR TITLE
Make doit(deep=False) of Add and Mul return evaluated result

### LINKED ISSUE #17280: Cleaning up Basic and subclasses for evaluate/doit
TL;DR I think that SymPy's model for Basic is messy in places and I think that there could be significant benefits in cleaning it up. Below are some ideas.

SymPy's Basic evaluation model broadly works as follows:

Basic subclasses define `__new__` which
1. Usually sympifies the arguments
2. Possibly raises if the arguments are of the wrong type/value
3. Possibly evaluates by simplifying the arguments (e.g. `re(x+I) -> re(x)`)
4. Possibly evaluates to return a different object (`Mul(2, 3) -> Integer(6)`)
5. Otherwise sends the arguments to `Basic.__new__` to create the object, set
args etc.

Evaluation is typically controlled in a few ways:
1. Some `__new__` functions don't evaluate when given `evaluate=False` giving en "unevaluated object".
2. Others don't evaluate by default and always return unevaluated.
3. Unevaluated objects can be evaluated by calling doit.
4. doit has kwargs like `deep` which can control whether doit is called recursively on args.

That all seems like the nice model but it isn't implemented consistently across the SymPy code base and presumably by downstream authors/users writing their own Basic subclasses.

## 1. Inconsistent support for evaluate kwarg

There are 351 `__new__` methods in SymPy:
```shell
$ git grep 'def __new__' | wc -l
     351
```
I (crudely) count 3 `__new__` methods that have an evaluate parameter and 124 that have `**kwargs`. That means that the majority of these classes canot accept an evaluate kwarg. Of those `__new__` methods that take `**kwargs` many do not respect any `evaluate` kwarg and will evaluate regardless.

Classes like Derivative/Integral/Subs that don't evaluate by default are inconsistent in accepting an evaluate kwarg. For example Derivative respects evaluate=True but Integral raises. There is also no programmatic way to tell whether a particular class accepts an evaluate kwarg or whether it would or would not evaluate by default.

I propose that all Basic types should accept an evaluate=True kwarg. As a kwarg it should default to None but True, False and None should be accepted.  In the case of None a class attribute (`cls.autodoit`?) should be checked. This allows Integral etc to disable evaluation by default at the class level. Subclasses should then not define `__new__` but something else like `_eval_new` that would be called by `Basic.__new__` (if evaluate=True - perhaps this should be `_eval_doit` - see below).

Simpification should be handled in `Basic.__new__` perhaps controlled by a
`cls.simpifyargs` class attribute for subclasses that specifically want to
accept unsimpified args. It might also be worthwhile to have an _eval_validate
method that can be used to validate the arguments (raising if necessary) when
evaluate=True. A sketch of how this could look:

```python
class Basic:

    sympifyargs = True
    autodoit = True

    def __new__(cls, *args, evaluate=None, **kwargs):
        if cls.simpifyargs:
            args = tuple(simpify(a) for a in args)

        cls._eval_validate(*args, **kwargs) # Possibly raises

        # Current code of Basic goes here
        obj = object.__new__(cls)
        obj._assumptions = cls.default_assumptions
        obj._mhash = None  # will be set by __hash__ method.
        obj._args = args  # all items in args must be Basic objects

        # Now evaluate to something else
        if evaluate is None:
            evaluate = cls.autodoit
        if evaluate:
            obj = obj._eval_new()

        return obj


class Tuple(Basic):

    sympifyargs = False

    @classmethod
    def _eval_validate(cls, t):
        if not isinstance(t, tuple):
            raise TypeError("A tuple is expected")


class ImageSet(Set):

    autodoit = False

    @classmethod
    def _eval_validate(cls, func, sym, set):
        if not isinstance(func, Function):
            raise TypeError
        elif not isinstance(sym, Symbol):
            raise TypeError
        elif not isinstance(set, Set):
            raise TypeError

    def _eval_new(self):
        if self.func.is_Identity:
            return self.set
        return self
```

## 2. Patchy implementations of doit.

There are 76 doit methods in SymPy. None have an explicit deep kwarg. Of the 62 that do take kwargs many do not check for any deep flag so it isn't generally possible to request a deep/shallow doit. Some doit methods are always deep and others are always shallow. Some objects have doit methods that don't undo the effect of evaluate=False so that `func(*args, evaluate=False).doit() != func(*args)`.

This ImageSet for example doesn't evaluate because it hasn't asked the Integral to doit:
```julia
In [1]: ImageSet(Lambda(x, x**2), Interval(0, 1)).doit()
Out[1]: [0, 1]

In [2]: ImageSet(Lambda(x, Integral(x, x)), Interval(0, 1)).doit(deep=True)
Out[2]:
⎧⌠                  ⎫
⎨⎮ x dx | x ∊ [0, 1]⎬
⎩⌡                  ⎭
```
I propose that doit should be defined as a method on Basic that handles options like deep. Subclasses should not define doit and should instead define `_eval_doit` that evaluates only the current object. If this was done consistently then it would be possible to do things like `expr.doit(Subs)`, `expr.doit(subexpr)`, or `expr.doit(exclude=Integral)`.

A sketch for doit:
```python
class Basic:
    def doit(self, **kwargs):
        obj = self
        if kwargs.get('deep', True):
            args = (a.doit(**kwargs) for a in obj.args)
            if args != obj.args:
                obj = type(obj)(*args, evaluate=False)
        exclude = kwargs.get('exclude', ())
        if not any(isinstance(obj, cls) for cls in exclude):
            return obj._eval_doit()
```

## 3. Refine and new assumptions

Currently there is no way to incorporate the new assumptions in evaluation except using global state as in `with assuming(assumptions)` which is not good as a general solution because globals are bad, not threadsafe etc. The code for refine is a parallel implementation of the code for normal evaluation. This means that refine cannot give results consistent with the old assumptions without reimplementing all of the evaluation code (`Mul.flatten` etc).

If evaluation was controlled in a more systematic way (as described above) then it would be possible to add assumptions arguments in key places in order to accommodate for the new assumptions in this setup. Ideally the code to evaluate an object would be the same as the code to refine the object except that additional assumptions get passed to ask internally. In particular `_eval_new` could have an assumptions argument and an `_eval_refine` method for Basic could be something like
```python
    def _eval_refine(self, assumptions):
        args = tuple(refine(a, assumptions) for a in self.args)
        return self._eval_new(*args, assumptions=assumptions)
```
The `_eval_new` can use `ask` internally instead of the old assumptions.

(This last part about refine is something I am less sure about. In particular I wonder about the fact that the new assumptions are built on top of Basic so I'm not sure how this would bootstrap. Maybe these changes belong at Expr rather than Basic but refine could be equally useful for non-Expr classes like sets.)

## 4. Backwards compatibility

I don't think it's reasonable to expect every author of a `doit/__new__` method to get all of the details right so it's important that a simpler subclassing interface is provided that naturally gets the details right.  Ideally I want to get to a state where implementing `doit/__new__` directly is discouraged (just like it is for Function subclasses which should define eval instead). However implementing `__new__` will still need to be supported for backwards compatiblity. It's also important that this work can be done incrementally rather than as a mega-patch.

I propose that the changes described above should be somehow opt-in and backwards compatible. This means in particular that anyone who previously wrote a class that defines `__new__` needs `Basic.__new__` to not trigger any evaluation. Maybe there should be some kind of version flag like `cls.basic_v2` or something to know whether `Basic.__new__` should be a no op. Perhaps testing for the presence of `_eval_new` would be sufficient for this.

If this was done in the right way I think it would be possible to make these changes incrementally across the SymPy codebase leading to a cleaner overall structure.

### LINKED ISSUE #18787: doit(deep=False): core vs matrices
In sympy core, `doit(deep=False)` does not evaluate unevaluated expression.
```python
>>> from sympy import Symbol, Add
>>> x = Symbol('x')
>>> expr1 = Add(x, -x, evaluate=False)
>>> expr1 # unevaluated
x - x
>>> expr1.doit(deep=False) # still unevaluated
x - x
```

However, in matrix module,  `doit(deep=False)` returns evaluated expression.
```python
>>> from sympy import MatrixSymbol, MatAdd
>>> A = MatrixSymbol('A', 2,2)
>>> expr2 = MatAdd(A, -A)
>>> expr2 # unevaluated
A -A
>>> expr2.doit(deep=False) #evaluated
ZeroMatrix(2, 2)
```

I think matrix module is doing it right. Shall we change core's `doit`?

### PR DESCRIPTION
**Pull request by mcpl-sympy**
#### References to other Issues or PRs

Partial fix for #17280
Partial fix for #18837

#### Brief description of what is fixed or changed

`doit(deep=False)` on `Add` and `Mul` now returns evaluated result.

```python
>>> from sympy import Add, Mul
>>> from sympy.abc import x
>>> Add(x, x, evaluate=False).doit(deep=False)
2*x
>>> Mul(x, x, evaluate=False).doit(deep=False)
x**2
```

#### Release Notes

- core
    - `doit(deep=False)` now evaluates `Add` and `Mul`.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
Unless you are intending to fix all of the problems mentioned in #17280 then this should not say "fixes #17280". You can see "partial fix for #17280" but that issue should not be closed when this is merged unless everything is resolved.

File: sympy/core/operations.py [hunk 1]
```diff
@@ -375,6 +375,13 @@ def make_args(cls, expr):
         else:
             return (sympify(expr),)
 
+    def doit(self, **hints):
+        if hints.get('deep', True):
+            terms = [term.doit(**hints) if isinstance(term, Basic) else term
+                                         for term in self.args]
```
Comment (oscarbenjamin) [inline sympy/core/operations.py]: When is term not Basic?

**Comment by oscarbenjamin** [comment]
I think it would be better if `doit` called `_eval_doit`. Then subclasses implement `_eval_doit` and `doit` should only be defined on `Basic` and should handle options like `deep=False` etc.

File: sympy/core/operations.py (hunk 1 above)
Comment (mcpl-sympy) [reply sympy/core/operations.py]: Fixed it. Thanks!

**Comment by mcpl-sympy** [comment]
@oscarbenjamin 
> I think it would be better if doit called _eval_doit. Then subclasses implement _eval_doit and doit should only be defined on Basic and should handle options like deep=False etc.  

That sounds nice. After modifying every `def __new__` in sympy, I will do it.

**Comment by asmeurer** [comment]
> Unless you are intending to fix all of the problems mentioned in #17280 then this should not say "fixes #17280". You can see "partial fix for #17280" but that issue should not be closed when this is merged unless everything is resolved.

It's important that you word it like that. Right now you have "This PR partly fixes #17280" which GitHub just sees as "fixes #17280" (note how it underlines the word "fixes").

**Comment by oscarbenjamin** [comment]
> After modifying every `def __new__` in sympy, I will do it.

I suggest to limit the scope of this work to something manageable at first. You should aim to get a single piece of work merged and then build on that. If you want to change everything at once then you will end up with an enormous PR and it likely won't get merged because it will be too hard for anyone else to review.

In #17372 I tried to fix these kinds of issues in just one package (sets) but found that there were a lot of badly written `Basic` subclasses that needed reworking. Just fixing doit for those classes that define doit is already a significant piece of work (there are 77 doit methods in sympy). Various bugs will probably be uncovered in the process and tests will need changing etc.

**Comment by mcpl-sympy** [comment]
@asmeurer  @oscarbenjamin 
Thanks. Then, in this PR, I will focus on changing `doit(deep=False)` on `Add` and `Mul`. I changed the PR description and made another Issue (#18837).  
Should this PR need more enhancement, please tell. Otherwise, this is ready to be merged.

**Comment by oscarbenjamin** [comment]
Looks good. Merging. Thanks for this!

**Comment by oscarbenjamin** [comment]
If you want to continue doing this fo all doit methods then I think it would be better to have `_eval_doit` rather than adding this kind of code in all 77 `doit` methods. Moving the logic for deep=False up to `Basic` and out of the subclasses will make it possible to have systematically controlled evaluation of `doit`.

**Comment by mcpl-sympy** [comment]
@oscarbenjamin
I tried, but modifying `Basic.doit` led to various test failures due to the inconsistencies. Thus I decided to make every classes take `evaluate` option first, then clean up `doit` methods. Until then, it will be better to override `doit` like this. (Removing these afterwards will take no effort)

**Pull request merged**